test/e2e/singlecluster/pod_test.go

```diff
@@ -223,6 +229,160 @@ var _ = ginkgo.Describe("Pod groups", func() {
+			})
+
+			defaultPriorityGroup := podtesting.MakePod("default-priority-group", ns.Name).
+				Image("gcr.io/k8s-staging-perf-tests/sleep:v0.1.0", []string{"-termination-code=1", "10min"}).
+				Queue(lq.Name).
+				Request(corev1.ResourceCPU, "2").
+				MakeGroup(2)
+			defaultGroupKey := client.ObjectKey{Namespace: ns.Name, Name: "default-priority-group"}
+
+			ginkgo.By("Default-priority group starts", func() {
+				for _, p := range defaultPriorityGroup {
+					gomega.Expect(k8sClient.Create(ctx, p.DeepCopy())).To(gomega.Succeed())
+				}
+				gomega.Eventually(func(g gomega.Gomega) {
+					for _, origPod := range defaultPriorityGroup {
+						var p corev1.Pod
+						gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(origPod), &p)).To(gomega.Succeed())
+						g.Expect(p.Spec.SchedulingGates).To(gomega.BeEmpty())
+					}
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			highPriorityGroup := podtesting.MakePod("high-priority-group", ns.Name).
+				Image("gcr.io/k8s-staging-perf-tests/sleep:v0.1.0", []string{"1ms"}).
+				Queue(lq.Name).
+				PriorityClass("high").
+				Request(corev1.ResourceCPU, "1").
+				MakeGroup(2)
+			highGroupKey := client.ObjectKey{Namespace: ns.Name, Name: "high-priority-group"}
+
+			ginkgo.By("Create the high-priority group", func() {
+				for _, p := range highPriorityGroup {
+					gomega.Expect(k8sClient.Create(ctx, p.DeepCopy())).To(gomega.Succeed())
+				}
+				gomega.Eventually(func(g gomega.Gomega) {
+					for _, origPod := range highPriorityGroup {
+						var p corev1.Pod
+						gomega.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(origPod), &p)).To(gomega.Succeed())
+						g.Expect(p.Spec.SchedulingGates).To(gomega.BeEmpty())
+					}
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+
+			ginkgo.By("The default priority workload is preempted", func() {
+				var updatedWorkload kueue.Workload
+				gomega.Expect(k8sClient.Get(ctx, defaultGroupKey, &updatedWorkload)).To(gomega.Succeed())
+				util.ExpectWorkloadsToBePreempted(ctx, k8sClient, &updatedWorkload)
+			})
+
+			replacementPods := make(map[types.NamespacedName]types.NamespacedName, len(defaultPriorityGroup))
+			ginkgo.By("Create replacement pods as soon as the default-priority pods are Failed", func() {
+				gomega.Eventually(func(g gomega.Gomega) int {
+					for _, origPod := range defaultPriorityGroup {
+						origKey := client.ObjectKeyFromObject(origPod)
+						if _, found := replacementPods[origKey]; !found {
+							var p corev1.Pod
+							gomega.Expect(k8sClient.Get(ctx, origKey, &p)).To(gomega.Succeed())
+							if p.Status.Phase == v1.PodFailed {
+								rep := origPod.DeepCopy()
+								// For replacement pods use args that let it complete fast.
+								rep.Name = "replacement-for-" + rep.Name
+								rep.Spec.Containers[0].Args = []string{"1ms"}
+								gomega.Expect(k8sClient.Create(ctx, rep)).To(gomega.Succeed())
```

Comment:
I wonder if there is potential for flakiness here, as Kueue might not have observed the Pod as Failed yet.

I'm not sure if events within a kind are ordered. In that case, Kueue might only see the replacement Pod after it has seen the other Pod as failed, in which case there wouldn't be flakiness.

Let's run the tests a few times.

Otherwise, we might have to implement the logic in which, instead of deleting excess Pods, they are left gated until there is space.

Reply:
I assume there is no issue with flakiness, because I run this in a look locally and for >1h and all attempts have passed. Also, all attempts on GH CI passed (6).